README.md

# edgeCounter

<!-- badges: start -->
[![R-CMD-check](https://github.com/yeyuan98/edgeCounter/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/yeyuan98/edgeCounter/actions/workflows/R-CMD-check.yaml)
[![BioC status](http://www.bioconductor.org/shields/build/release/bioc/edgeCounter.svg)](https://bioconductor.org/checkResults/release/bioc-LATEST/edgeCounter)
<!-- badges: end -->

`edgeCounter` is a R-package for counting edges (5'/3') of BAM alignment reads. 

For certain sequencing applications such as ATAC-seq, fragment edges, but not 
full-length fragments themselves, are physically meaningful. In these cases, simple 
genome coverage functions offered by widely adopted tools are not well-suited.

This package provides a streamlined solution for counting alignment read edges.

## Installation

You can install the development version of edgeCounter from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("yeyuan98/edgeCounter")
```

## Examples

Please refer to package vignettes for examples of `edgeCounter`.

## Issues

Please submit issues and/or pull requests as you see fit.


NEWS.md
# edgeCounter 0.99.1

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(edgeCounter)

test_check("edgeCounter")


tests/testthat/test-ranges.R
test_that("consensusGRanges() works with minimum.width defined", {
    # Testing by an artifical GRanges.
    #   NOTE - if the artifical GRanges described here changes,
    #     We need to save the `.rds` file again!
    gr <- makeGRanges(
        tibble::tribble(
            ~start, ~end, ~seqnames,
            1, 100, "chr2L",
            80, 120, "chr2L",
            200, 300, "chr2L",
            280, 320, "chr2L",
            360, 460, "chr2L",
            360, 400, "chr2L",
            100, 300, "chrX"
        )
    )
    expect_true({
        expect <- readRDS(test_path("fixtures", "consensusGRanges.example.rds"))
        all(expect == consensusGRanges(gr, minimum.width = 100))
    })
})


test_that("consensusGRanges() works with minimum.width NULL mode", {
    # Fully coded input and output here; no saved RDS file
    gr <- makeGRanges(
        tibble::tribble(
            ~start, ~end, ~seqnames,
            100, 200, "chr2L",
            150, 240, "chr2L",
            300, 500, "chr2L",
            490, 510, "chr2L",
            100, 200, "chrX",
            300, 400, "chrX"
        )
    )
    expect_true({
        expect <- makeGRanges(
            tibble::tribble(
                ~start, ~end, ~seqnames,
                100, 240, "chr2L",
                300, 510, "chr2L",
                100, 200, "chrX",
                300, 400, "chrX"
            )
        )
        all(expect == consensusGRanges(gr, minimum.width = NULL))
    })
})


test_that("readNarrowPeaks() works for single input", {
    # Testing by a single narrowPeak file provided.
    #   NOTE - the `.narrowPeak` and the `.rds` are linked.
    #     If we use a another `.narrowPeak`, we need to manually save `.rds` again!
    example.path <- test_path(
        "fixtures",
        "readNarrowPeaks.example.narrowPeak"
    )
    example.res.path <- test_path(
        "fixtures",
        "readNarrowPeaks.example.rds"
    )
    grs <- readNarrowPeaks(example.path)

    expect_equal(length(grs), 1)
    expect_equal(names(grs), example.path)
    expect_true(
        all(grs[[1]] == readRDS(example.res.path))
    )
})

test_that("readNarrowPeaks() works for multiple inputs with names", {
    # Testing by a single narrowPeak file provided, reading in two times.
    #   NOTE - the `.narrowPeak` and the `.rds` are linked.
    #     If we use a another `.narrowPeak`, we need to manually save `.rds` again!
    example.path <- test_path(
        "fixtures",
        "readNarrowPeaks.example.narrowPeak"
    )
    example.res.path <- test_path(
        "fixtures",
        "readNarrowPeaks.example.rds"
    )
    example.names <- c("f1", "f2")
    grs <- readNarrowPeaks(rep(example.path, times = 2),
        names = example.names
    )

    expect_equal(length(grs), 2)
    expect_true(
        all(names(grs) == example.names)
    )
    expect_true(
        all(grs[[1]] == readRDS(example.res.path))
    )
    expect_true(
        all(grs[[2]] == readRDS(example.res.path))
    )
})


tests/testthat/test-countEdges.R
test_that("bamEdges() working", {
    test.bam <- test_path("fixtures", "sample.bam")
    test.edges <- test_path("fixtures", "sample.bamEdges.rds")
    expect_true(
        all(bamEdges(test.bam) == readRDS(test.edges))
    )
})

test_that("countEdges() working", {
    test.bam <- test_path("fixtures", "sample.bam")
    test.range <- readRDS(test_path("fixtures", "sample.testGRanges.rds"))
    expect_equal(
        countEdges(test.range, test.bam)@elementMetadata@listData$edge.counts,
        c(2, 4, 0)
    )
})


R/utils.R
# Useful utility functions that satisfy the following criteria:
# 1. Wide applicability for testing/internal usage purposes
# OR
# 2. Useful for end-user to perform certain tasks

#' Simple wrapper for making GRanges
#'
#' @description `makeGRanges()` makes an artificial GRanges based
#' on user input. Useful for testing purposes and vignette building.
#'
#' @param df A tibble/data.frame object with at least two columns: start, end.
#' Optionally provide seqnames (otherwise will take default value)
#' Optionally provide strand (otherwise no strand information, i.e., '*')
#'
#' @param seqnames.default What value should seqnames take
#' if `is.null(df$seqnames)`.
#' This is restricted to length = 1 character vector.
#'
#' @examples
#' makeGRanges(
#'     tibble::tribble(
#'         ~start, ~end, ~seqnames,
#'         1, 100, "chr2L",
#'         80, 120, "chr2L",
#'         100, 300, "chrX"
#'     )
#' )
#'
#' makeGRanges(
#'     tibble::tribble(
#'         ~start, ~end,
#'         1, 100,
#'         80, 120,
#'         100, 300
#'     )
#' )
#'
#' @returns A GRanges object
#' @export
makeGRanges <- function(df, seqnames.default = "chr2L") {
    stopifnot(is.data.frame(df)) # tibble is also a data.frame
    stopifnot(
        !is.null(df$start),
        !is.null(df$end)
    )
    suppressWarnings(
        result <-
            GenomicRanges::GRanges(
                seqnames = ifelse(rep(is.null(df$seqnames), times = nrow(df)),
                    rep(seqnames.default, times = nrow(df)),
                    df$seqnames
                ),
                ranges = IRanges::IRanges(
                    start = df$start,
                    end = df$end
                ),
                strand = df$strand
            )
    )

    result
}


#' Break down a `GRanges` into a `GRangesList` of single ranges.
#'
#' @description `breakGRanges()` breaks a `GRanges` object into a `GRangesList`
#' where each list element is a single range (i.e., row)
#' in the original `GRanges`.
#'
#' @param gr Input `GRanges`
#' @returns A `GRangesList` object
breakGRanges <- function(gr) {
    res <-
        GenomicRanges::GRangesList(
            lapply(seq_along(gr), function(i) gr[i, ]),
            compress = TRUE
        )
    names(res) <- names(gr)
    res
}


#' Reading BED and BED-like file with minimum dependency
#'
#' @description `importBED()` read in a BED file as a GRanges object.
#'
#' @param path Path to the BED file. See details for format of the BED file.
#' @param extra.cols A named list for information on extra columns. It should
#' contain two elements `names` and `types` for name and type of the columns.
#'
#' @details
#' # Details
#'
#' ## BED file format
#'
#' The standard BED format supported by this function is three-column. Refer to
#' \href{https://genome.ucsc.edu/FAQ/FAQformat.html}{UCSC FAQ} for details.
#'
#' ## Extra columns
#'
#' Apart from the standard BED format, you can supply arbitrary extra columns
#' by providing names and types of the columns in the `extra.cols`
#' optional argument.
#'
#' The extra column names must be a character vector.
#'
#' Types must follow the shorthand types used by [readr::read_delim()].
#'
#' ## Strand column
#'
#' According to UCSC documentation, `strand` is an optional column
#' at position 6. Therefore, 6th column of the file, if existent,
#' must specify strand info. Supported strand spec characters are
#' `.` (i.e., no strand), `+` and `-`.
#'
#' @returns A GRanges object
#' @examples
#' bed.path <- system.file("extdata", "importBED.example.bed",
#'     package = "edgeCounter"
#' )
#' importBED(bed.path)
#' @export
importBED <- function(path, extra.cols = NULL) {
    # Sanity checks of the extra column specs if provided
    if (!is.null(extra.cols)) {
        stopifnot(
            is.character(extra.cols$names),
            is.character(extra.cols$types)
        )
    }

    # Get the full column specs
    names <- c(
        c("seqnames", "start", "end"), # BED default
        extra.cols$names
    )
    types <- paste0(
        "cii", # BED default
        extra.cols$types
    )

    # Read in the BED file
    df <- readr::read_delim(
        file = path, delim = "\t",
        col_names = names, col_types = types
    )

    # Add strand information if provided
    if (ncol(df) >= 6) {
        # Extract strand
        strands <- df[[6]]
        # Sanity check
        stopifnot(
            is.character(strands),
            all(strands %in% c(".", "+", "-"))
        )
        # Change `.` (UCSC) to `*` (GenomicRanges)
        strands[strands == "."] <- "*"
        # Write strand information and make sure it is named `strand`.
        #   This `strand` name is used by `makeGRanges` function.
        df[[6]] <- strands
        names(df)[6] <- "strand"
    }

    # Convert into GRanges
    gr <- makeGRanges(df)

    # Adding the metadata columns
    df$start <- NULL
    df$end <- NULL
    df$seqnames <- NULL
    suppressWarnings({
        df$strand <- NULL # tibble warns if `strand` column is not existent.
    })
    GenomicRanges::mcols(gr) <- df

    gr
}


#' Convert rowRanges to BED-format
#'
#' @description `rowRangesToBED()` takes in a `RangedSummarizedExperiment`
#' and converts its `rowRanges()` GRangesList into BED-formatted list.
#'
#' @param edgeExp A `RangedSummarizedExperiment` object
#'
#' @returns A list of BED-formatted rowRanges. See details.
#'
#' @details
#' # Details
#'
#' Returned list is essentially a data.frame with the following columns.
#' `x=y` below means that `x` is the list element in the returned list,
#' and `y` is the `GRanges` equivalent.
#'
#' ```
#' chrom = seqnames; Each GRanges MUST contain only one unique seqnames.
#' chromStart = min(start())
#' chromEnd = max(end())
#' name = name of the GRanges defined in rowRanges list
#' strand = Always set to '.'
#' thickStart = min(start())
#' thickEnd = max(end())
#' itemRgb = Always set to '255,0,0'
#' blockCount = length()
#' blockSizes = width()  # comma-separated list
#' blockStarts = start() - min(start())  # comma-separated list
#' ```
#'
#' ## Performance note
#'
#' Current implementation is relatively slow due to the `vapply`s
#' involving `IRanges::RleList`.
rowRangesToBED <- function(edgeExp) {
    # Sanity check
    stopifnot(inherits(edgeExp, "RangedSummarizedExperiment"))
    seqnames <- SummarizedExperiment::seqnames(edgeExp)
    seqnames.rn <- vapply(seqnames, S4Vectors::nrun, 0)
    #   All ranges must have only ONE run for seqnames
    stopifnot(all(seqnames.rn == 1))
    # Get rowRanges
    rr <- SummarizedExperiment::rowRanges(edgeExp)
    # Parse names
    rr.name <- names(rr)
    if (is.null(rr.name)) rr.name <- rep(NA_character_, times = length(rr))
    # Parse starts and ends
    starts <- min(GenomicRanges::start(rr))
    ends <- max(GenomicRanges::end(rr))
    # Parse blockSizes
    blocksizes <- vapply(rr, function(gr) {
        paste(GenomicRanges::width(gr), collapse = ",")
    }, "")
    # Parse blockStarts
    blockstarts <- vapply(rr, function(gr) {
        abs.start <- min(GenomicRanges::start(gr))
        rel.starts <- GenomicRanges::start(gr) - abs.start
        paste(rel.starts, collapse = ",")
    }, "")
    list(
        chrom = vapply(seqnames, function(sn) as.character(sn)[1], ""),
        chromStart = starts,
        chromEnd = ends,
        name = rr.name,
        strand = rep(".", times = length(rr)),
        thickStart = starts,
        thickEnd = ends,
        itemRgb = rep("255,0,0", times = length(rr)),
        blockCount = vapply(rr, length, 0),
        blockSizes = blocksizes,
        blockStarts = blockstarts
    )
}


#' Construct UCSC-format BED track line with score display
#'
#' @description `getBEDtrackLine()` constructs "track line" based on
#' [UCSC guide](https://genome.ucsc.edu/FAQ/FAQformat.html#format1).
#' It tries to turn on score display based on settings suggested in
#' [IGV help form](https://groups.google.com/g/igv-help/c/IyBzO5_j5E8).
#'
#' @param name Name of the track
#' @param score Scores of the track, for setting up score display limits.
#' Although there is no upper limit, scores MUST be >= 0.
#'
#' @returns A string of "track line" without newline character.
getBEDtrackLine <- function(name, score) {
    # Sanity check
    stopifnot(is.character(name))
    stopifnot(is.numeric(score), all(score >= 0))
    # Compute display limits
    ulim <- max(score)
    llim <- min(score)
    # Return track line
    paste("track",
        paste0("name=", name),
        "useScore=1",
        paste0("viewLimits=", llim, ":", ulim),
        "color=255,0,0",
        sep = "\t"
    )
}


R/countEdges.R
# Main functions for counting edges

#' Get edges of BAM alignment reads
#'
#' @description
#' `bamEdges()` reads a BAM file and returns the 5' AND 3' ends of EACH
#' paired-end read PAIR as a `GRanges` object.
#'
#' @param bam.path Path to the BAM file. The BAM file is presumed to contain
#' read PAIRs.
#' @returns A `GRanges` object representing edges of BAM reads.
#'
#' @details
#' # Use cases
#'
#' * For ATAC sequencing, read edges are roughly transposition sites of the Tn5
#' transposase. To get the exact base-pair-precision transposition sites
#' you need to manually shift the read start/end positions.
#'
#' * Other cases?
#'
#' @examples
#' bam.path <- system.file("extdata", "sample.near.Gapdh1.bam",
#'     package = "edgeCounter"
#' )
#' edges <- bamEdges(bam.path) # A GRanges containing read edges
#' @export
bamEdges <- function(bam.path) {
    # Read in BAM file; only take necessary fields
    scanParam <- Rsamtools::ScanBamParam(
        what = c("rname", "pos", "isize"),
        # For a pair, only take one mate
        flag = Rsamtools::scanBamFlag(isMinusStrand = FALSE)
    )
    bam <- Rsamtools::scanBam(bam.path, param = scanParam)[[1]]

    # Get GRange of the 5' and 3' ends for each read pair
    # each read PAIR will be converted to TWO entries in the GRange
    ends <- c(bam$pos, bam$pos + abs(bam$isize))
    # each entry is of width=1 reflecting the 5' and 3' ends of the PAIR
    GenomicRanges::GRanges(
        seqnames = c(bam$rname, bam$rname),
        ranges = IRanges::IRanges(
            start = ends,
            end = ends
        ),
        strand = NULL,
        seqinfo = NULL
    )
}


#' Count number of BAM read edges for each range of a `GRanges/GRangesList`.
#'
#' @description `countEdgesCached()` counts number of BAM read pair edges for
#' each range of a given `GRanges` object. Compared to
#' the exported `countEdges()`, `countEdgesCached()` requires providing
#' the pre-extracted edges using `bamEdges()`.
#'
#'
#' @param ranges A GRanges/GRangesList object containing the ranges to consider
#' @param edges A GRanges containing the edges. Use `bamEdges()` to get this.
#' @returns The input GRanges/GRangesList object with
#' an appended metadata column `edge.counts`.
#'
#' @seealso
#' For more detailed discussion on "edges"
#' of a BAM alignment file, see [bamEdges()].
#' Consult [GenomicRanges::GRanges] about `GRanges` objects.
countEdgesCached <- function(ranges, edges) {
    counts <- GenomicRanges::countOverlaps(ranges, edges, type = "any")
    GenomicRanges::mcols(ranges)[["edge.counts"]] <- counts

    ranges
}


#' Count number of BAM read edges for each range of a `GRanges/GRangesList`.
#'
#' @description `countEdges()` counts number of BAM read pair edges for each
#' range of a given `GRanges` object.
#'
#'
#' @param ranges A GRanges/GRangesList object containing the ranges to consider
#' @param bam.path Path to the BAM file.
#' @returns The input GRanges/GRangesList object with
#' an appended metadata column `edge.counts`
#'
#' @seealso
#' For more detailed discussion on "edges"
#' of a BAM alignment file, see [bamEdges()].
#' Consult [GenomicRanges::GRanges] about `GRanges` objects.
#' This function calls internal function [countEdgesCached()].
#'
#' @examples
#' bam.path <- system.file("extdata", "sample.near.Gapdh1.bam",
#'     package = "edgeCounter"
#' )
#' ranges <- makeGRanges(
#'     tibble::tribble(
#'         ~start, ~end, ~seqnames,
#'         7791000, 7794000, "chr2R", # inclusive of Gapdh1 locus
#'         80, 120, "chr2L", # random range
#'         100, 300, "chrX" # random range
#'     )
#' )
#' ranges <- countEdges(ranges, bam.path) # Adds a metadata column of counts
#' @export
countEdges <- function(ranges, bam.path) {
    edges <- bamEdges(bam.path)
    countEdgesCached(ranges, edges)
}


R/ecParallel.R
# Parallel computing supporting functions
#   Uses global options, refer to `edgeCounterOptions`


#' Set parallel computing options of edgeCounter
#'
#' @description `ecParallel()` sets global parallel computing options.
#'
#' @param PSOCK.names `name` of [parallel::makePSOCKcluster].
#' If `NULL` is provided, parallel computing feature will be turned off.
#'
#' @returns None.
#' @examples
#' ecParallel(6) # 6 workers, local machine
#' getOption("ecOptions")
#' ecParallel(NULL) # turn off PSOCK parallelization
#' getOption("ecOptions")
#' @export
ecParallel <- function(PSOCK.names) {
    # Sanity checking
    valid <- function(x) is.null(x) | ecOptions.valid()[["PSOCK.names"]](x)
    stopifnot(valid(PSOCK.names))

    # Set parallel options
    if (is.null(PSOCK.names)) {
        ecOptions(parallel = FALSE)
    } else {
        ecOptions(parallel = TRUE, PSOCK.names = PSOCK.names)
    }

    invisible()
}


#' Verify if an object is list-like
#'
#' @description `is.list.like()` checks if an object is 'list-like'.
#'
#' @param obj A single object
#'
#' @returns Boolean where TRUE means the object is 'list-like'.
#'
#' @details
#'
#' A 'list-like' object is anything that has the following operators:
#'
#' * `[[]]` double bracket subsetting, which should
#' return an object in the list. Integer subsetting must be supported.
#' * `length` which should return length of the list.
#'
#' @examples
#' gr <- makeGRanges(
#'     tibble::tribble(
#'         ~start, ~end, ~seqnames,
#'         1, 100, "chr2L",
#'         80, 120, "chr2L",
#'         100, 300, "chrX"
#'     )
#' )
#'
#' is.list.like(gr) # FALSE
#' is.list.like(GenomicRanges::GRangesList(list(gr, gr))) # TRUE
#'
#' @export
is.list.like <- function(obj) {
    tryCatch(
        {
            length(obj)
            obj[[length(obj)]]
            return(TRUE)
        },
        error = function(e) {
            return(FALSE)
        }
    )
}


#' Parallel execution of any function with arbitrary arguments.
#'
#' @description `ecParallelFunc()` runs any function with arbitrary arguments
#' in a parallel manner specified by `ecOptions()`.
#'
#' @param func Function to run, which must take arguments specified later.
#' @param benchmark If TRUE will message execution time. The time measured
#' includes all overheads such as PSOCK cluster initialization.
#' @param ... Arbitrary arguments of the function. All arguments must be
#' list-like' objects (see details).
#'
#' @returns A list of the execution results.
#' @details
#' # Arguments supplied
#'
#' `ecParallelFunc()` takes arbitrary arguments supplied as list-like objects
#' and feed consecutive combinations of the arguments.
#'
#' ## 'list-like' objects
#'
#' A list-like object must support the following operators.
#'
#' * `[[]]` double bracket subsetting, return an individual object in the list.
#' * `length` length accessor, return length of the list.
#'
#' # Behavior depending on `ecOptions`
#'
#' [ecOptions()] describes options of this package
#' including parallel computation.
#' If `parallel == FALSE`, `ecParallelFunc()` will be running
#' the combinations in serial. If `parallel == TRUE`,
#' `ecParallelFunc()` will initialize PSOCK cluster, execute computation,
#' stop the PSOCK cluster, and then return the results.
#'
#' @examples
#' ecParallelFunc(function(x, y) x + y, x = c(1, 2, 3), y = list(5))
#' @export
ecParallelFunc <- function(func, benchmark = FALSE, ...) {
    # If `benchmark` get time
    if (benchmark) time <- Sys.time()
    # Get the arguments
    args <- list(...)
    # Sanity checking
    stopifnot(is.function(func)) #   Must provide a function
    stopifnot(all(vapply(args, is.list.like, TRUE))) #   list-like arguments
    #   Argument list lengths must be the same or 1
    arg.lens <- vapply(args, length, 1)
    arg.lens <- dimnames(table(arg.lens))[[1]]
    arg.lens <- arg.lens[arg.lens != "1"]
    stopifnot(length(arg.lens) %in% c(0, 1))
    # Decide on parallel computation
    par <- getOption("ecOptions")$parallel
    if (par) {
        # Start cluster
        par.names <- getOption("ecOptions")$PSOCK.names
        par.packages <- getOption("ecOptions")$parallel.packages
        cl <- parallel::makePSOCKcluster(par.names)
        # Load packages
        parallel::clusterCall(cl,
            function(packages) {
                lapply(packages, function(x) do.call(require, list(x)))
            },
            packages = par.packages
        )
        # Load function for computation
        parallel::clusterExport(cl, "func", envir = environment())
        pMApply <- function(...) {
            parallel::clusterMap(
                cl = cl, fun = func,
                .scheduling = "static", ...
            )
        }
    } else {
        pMApply <- function(...) mapply(FUN = func, SIMPLIFY = FALSE, ...)
    }
    # Apply function
    results <- pMApply(...)
    # Clean up
    if (par) {
        parallel::stopCluster(cl)
    }
    # If `benchmark` message elapsed time
    if (benchmark) {
        message(
            "Elapsed time ",
            format(round(difftime(Sys.time(), time, units = "secs"),
                digits = 1
            ))
        )
    }

    results
}


R/edgeCounterOptions.R
# Global options of edgeCounter package, used by various functions
#   The following aspects are considered currently:
#     parallel computation support (by PSOCK cluster)
#       `parallel` - BOOLEAN
#       `PSOCK.names` - integer or character, see `parallel::makeCluster`
#       The package always stop PSOCK clusters on exit of supported functions.
#   Check out zzz.R for initialization of the options.


#' Validation of global options
#'
#' @description `ecOptions.valid()` is internally used for global option
#' parameters sanity validation.
#'
#' @returns A named list where names are names of available global options and
#' values are functions for validating the option.
#' Validation functions must take in a single argument and return Boolean.
ecOptions.valid <- function() {
    list(
        parallel = function(x) is.logical(x) & length(x) == 1,
        PSOCK.names = function(x) is.numeric(x) | is.character(x),
        parallel.packages = function(x) is.character(x)
    )
}

#' Defaults of global options
#'
#' @description `ecOptions.default()` is internally used for global option
#' defaults
#'
#' @returns A named list of the default values for the global options.
ecOptions.default <- function() {
    list(
        parallel = FALSE,
        PSOCK.names = NA,
        parallel.packages = "edgeCounter"
    )
}


#' Global options for the edgeCounter package
#'
#' @description `ecOptions()` sets edgeCounter package global options.
#'
#' @param ... Parameter name-value pairs supported by the package. See details
#' for a list of global parameters. If no parameter is provided, do not change
#' current options.
#'
#' @returns Up-to-date edgeCounter options.
#'
#' @details
#'
#' Global options are accessible by `getOption("ecOptions")`.
#' # Supported options
#'
#' * `parallel`: a Boolean specifying whether to enable parallel computation.
#'
#' * `PSOCK.names`: PSOCK cluster `names` parameter.
#' See [parallel::makePSOCKcluster]. Not used if `parallel` is false.
#'
#' * `parallel.packages`: Which packages to load in PSOCK cluster nodes.
#'
#' ## Parallel computation
#'
#' This package always use PSOCK cluster as supported by the `parallel` package
#' for parallel computation. A 'diligent' approach for initialization
#' and clean-up of the PSOCK cluster is used, meaning that the PSOCK cluster is
#' initialized only when functions supporting parallel computation is called,
#' and the cluster is stopped whenever the called function exits.
#' @examples
#' ecOptions() # Returns current option
#' ecOptions(parallel = FALSE) # Turns off parallel computation
#' @export
ecOptions <- function(...) {
    # INTERNAL NOTE - WHEN INITIALIZING THE PACKAGE, FEED NO ARGUMENT IS FINE.

    # Fetch arguments into a list
    args <- list(...)

    # If arguments are provided, we validate it
    if (length(args) != 0) {
        # Sanity check
        supported <- ecOptions.valid()
        #   All arguments must be supported
        stopifnot(all(names(args) %in% names(supported)))
        #   All argument values must be valid
        stopifnot(all(vapply(
            names(args),
            function(n) supported[[n]](args[[n]]),
            TRUE
        )))
    }

    # Set the global options
    #   Fetch current options
    opt <- getOption("ecOptions", default = list())
    #   Change options
    for (n in names(args)) opt[[n]] <- args[[n]]
    #   Fill undefined options with defaults
    def <- ecOptions.default()
    for (n in names(def)) if (!(n %in% names(opt))) opt[[n]] <- def[[n]]
    #   Update options
    options(ecOptions = opt)

    opt
}


R/ranges.R
# How to establish the intended GRanges object for counting?
#   The following methods are suggested:
#     1. Read MACS2 narrowPeaks or other BED-like files
#     2. Genomic features from packages, e.g., `GenomicFeatures::TxDb`.
#   This package provides a set of helper functions to deal with #1.
#   To process #2 data, refer to `inter-range-methods` of `GenomicRanges`.

#' Read in MACS2 narrowPeaks from one/multiple files into GRanges objects
#'
#' @description `readNarrowPeaks()` reads in MACS2 narrowPeaks from files
#' into a list of GRanges object(s).
#'
#'
#' @param paths A character vector of MACS2 narrowPeak file paths
#' @param names (optional) A character vector for names of the GRanges
#' @returns A named list of GRanges objects. Names are the input paths.
#' If `names` is provided, `names` is used
#' as list names instead of input paths.
#'
#' @examples
#' np.path <- system.file("extdata", "readNarrowPeaks.example.narrowPeak",
#'     package = "edgeCounter"
#' )
#' readNarrowPeaks(np.path, names = "sample") # Returns a list of GRanges
#' @export
readNarrowPeaks <- function(paths, names = NULL) {
    # Extra columns of narrowPeak file
    extraCols <- list(
        names = c(
            "name", "score", "strand", "foldChange",
            "-log10Pvalue", "-log10Qvalue", "summitPos"
        ),
        types = "cicdddi"
    )

    # Read in each narrowPeak file using the `importBED` of this package
    grs <- lapply(paths, function(path) {
        importBED(path = path, extra.cols = extraCols)
    })

    # Assign list names
    if (is.null(names)) {
        names(grs) <- paths
    } else {
        stopifnot(length(names) == length(paths))
        names(grs) <- names
    }

    grs
}

#' Compute 'consensus' GRanges from a list of GRanges objects
#'
#' @description `consensusGRanges()` computes consensus GRanges from a list of
#' GRanges objects. Definition of 'consensus' is explained in Details section.
#'
#'
#' @param grs A single GRanges or a list of GRanges, could be length 1 or more.
#' @param minimum.width Minimum width of the consensus ranges. Could be either
#' positive integer or NULL (which will trigger two computation modes, see
#' detail).
#' @returns A GRanges object consisting the consensus ranges.
#'
#' @details
#' # Details
#'
#' This section describes behavior of this function in detail. Two modes are
#' supported by this function, which are defined by value of `minimum.width`.
#'
#' ## Consensus algorithm (when `minimum.width` is NOT `NULL`)
#'
#' The following steps are adopted to compute a single `GRanges` object from
#' a list of input `GRanges`.
#' Inter-range methods of the `GenomicRanges` package are used for computation.
#'
#' 1. Concatenate all ranges from all GRanges in list into a single GRanges.
#' 2. Get a collection of non-overlapping ranges by `disjoin`.
#' 3. Only retain non-overlapping ranges that overlaps with at least two ranges
#' from the original concatenated GRanges.
#' 4. Merge connecting (i.e., distance is zero bp) ranges by `reduce`.
#' 5. Extend ranges to >= `minimum.width`bp in length (or keep unchanged).
#' 6. Merge connecting ranges by `reduce`.
#' 7. Those ranges are considered to be 'consensus'.
#'
#' ### Minimum width of the consensus ranges
#'
#' When `minimum.width` is a defined integer, the consensus ranges computed by
#' the algorithm above is enforced to be at least of `minimum.width` in width.
#' If ranges are not wide enough, they are extended to `minimum.width` width by
#' `GenomicRanges::resize(..., fix = "center")`.
#'
#' ## Consensus algorithm (when `minimum.width` is `NULL`)
#'
#' When minimum.width is set to be `NULL`, the function tries to get "widest"
#' and "most" possible consensus ranges. The following steps are adopted:
#'
#' 1. Concatenate all ranges from all GRanges in list into a single GRanges.
#' 2. Get a collection of non-overlapping ranges by `disjoin`.
#' 3. Merge connecting (i.e., distance is zero bp) ranges by `reduce`.
#' 4. Those ranges are considered to be 'consensus'.
#'
#' Therefore, under this mode even ranges that occur only once will be reported
#' as a "consensus range".
#'
#' @seealso
#' The following inter-range methods from the `GenomicRanges` package are used:
#'
#' * [GenomicRanges::disjoin()]
#' * [GenomicRanges::reduce()]
#' * [GenomicRanges::countOverlaps()]
#'
#' For `countOverlaps`, this functions always adopt parameter `type = "any"`.
#'
#' @examples
#' grs <- makeGRanges(
#'     tibble::tribble(
#'         ~start, ~end, ~seqnames,
#'         100, 200, "chr2L",
#'         150, 250, "chr2L",
#'         200, 400, "chrX",
#'         250, 500, "chrX"
#'     )
#' )
#' consensusGRanges(grs) # With default, a minimum width is enforced
#'
#' grs <- makeGRanges(
#'     tibble::tribble(
#'         ~start, ~end, ~seqnames,
#'         100, 240, "chr2L",
#'         300, 510, "chr2L",
#'         100, 200, "chrX"
#'     )
#' )
#' consensusGRanges(grs, minimum.width = NULL)
#'
#' @export
consensusGRanges <- function(grs, minimum.width = 100) {
    # Sanity check of minimum.width
    stopifnot((is.numeric(minimum.width) & minimum.width > 0) |
        is.null(minimum.width))
    # Convert into a single GRanges of peaks and strip metadata
    if (is.list(grs)) grs <- Reduce(c, grs)
    stopifnot(inherits(grs, "GRanges"))
    GenomicRanges::mcols(grs) <- NULL

    # Get disjoin set - those are potential consensus peaks
    grs.dj <- GenomicRanges::disjoin(grs)

    if (!is.null(minimum.width)) {
        # Disjoin ranges that overlap >1 times with the original
        grs.dj.count <- GenomicRanges::countOverlaps(grs.dj, grs, type = "any")
        consensus <- grs.dj[grs.dj.count > 1]
        consensus <- GenomicRanges::reduce(consensus)

        # For those width < minimum.width adjust their width with center fixed
        consensus.small.set <- GenomicRanges::width(consensus) < minimum.width
        consensus.small <- consensus[consensus.small.set]
        consensus.small <- GenomicRanges::resize(
            x = consensus.small,
            width = minimum.width,
            fix = "center"
        )
        consensus[consensus.small.set] <- consensus.small

        # Reduce the ranges
        consensus <- GenomicRanges::reduce(consensus, drop.empty.ranges = TRUE)
    } else {
        consensus <- GenomicRanges::reduce(grs.dj, drop.empty.ranges = TRUE)
    }

    consensus
}
